h2o-admissibleml/src/main/java/hex/Infogram/Infogram.java

```diff
@@ -29,26 +26,111 @@
   String[] _modelDescription; // describe each model in terms of predictors used
-  int _numModels;             // number of models to build
-  double[] _cmi;              // store conditional mutual information
-  double[] _cmiRaw;           // conditional mutual information before normalization to 0 and 1
+  int _numModels; // number of models to build
+  double[] _cmi;  // store conditional mutual information
+  double[] _cmiValid;
+  double[] _cmiCV;
+  double[] _cmiRaw;  // raw conditional mutual information
+  double[] _cmiRawValid;  // raw conditional mutual information from validation frame
+  double[] _cmiRawCV;
+  String[] _columnsCV;
   TwoDimTable _varImp;
-  int _numPredictors;         // number of predictors to consider in training dataset
-  Key<Frame> _cmiRelKey;      // store frame key containing column, admissibe, admissible_index, relevance, cmi, cmi_raw
+  int _numPredictors; // number of predictors in training dataset
+  Key<Frame> _cmiRelKey;
+  Key<Frame> _cmiRelKeyValid;
+  Key<Frame> _cmiRelKeyCV;
   List<Key<Frame>> _generatedFrameKeys; // keep track of all keys generated
-
+  boolean _cvDone = false;  // on when we are inside cv
+  private transient InfogramModel _model;
+  long _validNonZeroNumRows;
+  int _nFoldOrig = 0;
+  Model.Parameters.FoldAssignmentScheme _foldAssignmentOrig = null;
+  String _foldColumnOrig = null;
+  
   public Infogram(boolean startup_once) { super(new hex.Infogram.InfogramModel.InfogramParameters(), startup_once);}
 
   public Infogram(hex.Infogram.InfogramModel.InfogramParameters parms) {
     super(parms);
     init(false);
   }
 
+  public Infogram(hex.Infogram.InfogramModel.InfogramParameters parms, Key<hex.Infogram.InfogramModel> key) {
+    super(parms, key);
+    init(false);
+  }
+  
   @Override
   protected Driver trainModelImpl() {
     return new InfogramDriver();
   }
 
+  @Override
+  protected int nModelsInParallel(int folds) {
+    return nModelsInParallel(folds,2);
+  }
+
+  /***
+   * This is called before cross-validation is carried out
+   */
+  @Override
+  public void computeCrossValidation() {
+    info("cross-validation", "cross-validation infogram information is stored in frame with key" +
+            " labeled as relevance_cmi_key_cv and the admissible features in admissible_features_cv.");
+    if (error_count() > 0) {
+      throw H2OModelBuilderIllegalArgumentException.makeFromBuilder(Infogram.this);
+    }
+    super.computeCrossValidation();
+  }
+
+  // find the best alpha/lambda values used to build the main model moving forward by looking at the devianceValid
+  @Override
+  public void cv_computeAndSetOptimalParameters(ModelBuilder[] cvModelBuilders) {
+    int nBuilders = cvModelBuilders.length;
+    double[][] cmiRaw = new double[nBuilders][];
+    List<List<String>> columns = new ArrayList<>();
+    long[] nObs = new long[nBuilders];
+    for (int i = 1; i < cvModelBuilders.length; i++) {  // run cv for each lambda value
```

Comment:
@wendycwong, we maybe confuse you a little bit with @koniecsveta. I thought that the ++i increase the index before iteration, but it is wrong. So I think your old code was correct. @koniecsveta just wanted to keep the same style of all the `for` cycles, because it is more clear to everyone then. I am not sure if your code is working correctly now. Do you really need to start from 1? I thought it is the goal, but I may be confused you. Is it working correctly now? 

If it is wrong now, my big apologies I confused you!

Reply:
@maurever I will change it back.  No worries.  Thank you so much for all the help you and zuzana have been given me.  I will change it back and it should be fine.